# MATLAB and Linear Algebra
As mentioned in the previous section, SPM is a software package that is built *on top* of the MATLAB scientific computing environment. Although possible to use SPM without any further knowledge of MATLAB, there is an argument to be made for being at least a *bit* familiar with the MATLAB language. Unfortunately, we do not have time on this course for in-depth training on computer programming. However, we can start using MATLAB as a way of understanding some of the mathematical theory that we will need going forward. This will do double-duty as both a way of getting familiar with MATLAB and a way of learning some mathematical concepts in a more interactive fashion. 

### The MATLAB Prompt
In order to communicate with MATLAB, we use the MATLAB prompt. The prompt is signified in the Command Window by the two angled brackets `>>`. These brackets indicate where we can type commands for MATLAB to interpret. We type something at the prompt, hit <kbd>ENTER</kbd> and then whatever we typed gets sent to the MATLAB interpreter. If MATLAB understands what we typed, it will return the answer. If MATLAB does not understand what we typed, it will return an error.

As an example, if we wanted to use MATLAB as a highly over-engineered calculator, we could ask what $3 \times 4$ is by typing

In [1]:
3 * 4

ans = 12

However, if we accidentally typed `$` instead of `*`, we would get

In [4]:
3 $ 4

Invalid text character. Check for unsupported symbol, invisible character, or pasting of non-ASCII characters.

In this instance, MATLAB has return an error because it does not undestand what `$` means in this context. 

`````{admonition} Tip
:class: tip
When faced with an error, part of the skill of using a programming language is working out why a certain command failed. Even for seasoned programmers, errors are common. Errors will decrease with experience, but they will never disappear. The trick is not to panic, but to read what the error says slowly and carefully. Although tempting to assume the problem is with MATLAB, 9/10 it is a user error. As such, you need to try and develop some skills in discerning what seemingly cryptic error messages mean so that you do not get stuck. This comes with time and experience, but is a valuable skill to develop.
`````

## Variables
Before getting into some of the key linear algebra concepts, we first need to understand some programming fundamentals. One of the most basic characteristics of any programming language is the ability to store things in memory for later use. We do this by using *variables*, which you can think of as boxes with names that we can put something inside to keep for later. As an example, the code below shows the declaration of a variable called `a` and assigning it the number 5.

In [86]:
a = 5;

Notice here that we have used a semi-colon `;` at the end of the line to suppress MATLAB printing any output. In declaring the variable `a`, we have created a box that we have called `a` that contains the number `5`. Now, every time we type `a`, MATLAB will open this box and use whatever is inside it. As such, if we just type `a` we get 

In [88]:
a

a = 5

but if we do something with the returned value we get

In [90]:
a + 7

ans = 12

`````{admonition} Tip
:class: tip
Whenever a value is not explicitly assigned to a variable, MATLAB will automatically assign it to a default variable called `ans` (for *answer*).
`````

Similarly, we could declare another variable with the name `b`

In [92]:
b = 10;

and then add `a` and `b` together

In [94]:
a + b

ans = 15

Although these examples are quite trivial, the power of variables means that we can create code where we do not know the values of `a` and `b`. Perhaps their value depends upon the data we are using? Or perhaps it depends on some other factor that may change each time the code runs? Either way, it does not really matter what value they have if we know that we want to add them together. Thus we can write generic code that can be used irrespective of what the inputs will be.

`````{admonition} Tip
:class: tip
There are rules about naming variables. A variable name cannot start with a number, it cannot contain spaces and the only special character you can use is an underscore. You also need to be aware that MATLAB is case-sensitive, meaning that `variable_a` and `VARIABLE_A` will be interpreted as different entities. MATLAB does not know what you meant to type, only what you *did* type. In addition, remember that a variable can only hold one thing at a time and if you re-assign a variable then the original value will be lost.
`````

## Vectors and Matrices
Now that we have the idea of variables in mind, we can move on to understanding some basic principles of linear algebra. Just to reiterate, these are concepts that we will need going forward on the course. As such, it is important to gain some familiarity at this point, even though the context will be missing. Just try and get a feel for some of these ideas and have a go at feeding the code into MATLAB. We will start to see all of this in action as the course progresses, so just do your best to try and understand it even if the relevance is not immediately apparent.  

To begin with, the most fundamental entities in linear algebra are *vectors* and *matrices*. We will be using these a lot on this course, so it is important to both understand what a vector or matrix is, as well as understanding how they can be manipulated.

### Vectors
A vector is simply a list of numbers. For instance

$$
\mathbf{a} = \begin{bmatrix}3 & 4 & 5\end{bmatrix}.
$$

Here, $\mathbf{a}$ is a 3-dimensional row vector with elements 3, 4 and 5. We can refer to the individual elements using *subscripts*. For instance, $a_{1} = 3$, $a_{2} = 4$ and so on. Typically, a vector is written as a **bold** lower-case letter. A vector can also be defined as a *column* of numbers, for instance

$$
\mathbf{b} = 
\begin{bmatrix}
    6 \\  
    7 \\ 
    8 \\
\end{bmatrix}.
$$

Here, $\mathbf{b}$ is a 3-dimensional column-vector with elements $b_{1} = 6$, $b_{2} = 7$ and $b_{3} = 8$. Although the orientation of a vector may seem like a trivial detail, we will soon see that this is very important.

In MATLAB, a vector is defined using a list of numbers between square brackets. For instance, we can define the vector $\mathbf{a}$ in MATLAB as

In [98]:
a = [3 4 5]

a = 1x3 double
     3     4     5

which tells us that `a` is a $1 \times 3$ vector of [double-precision](https://en.wikipedia.org/wiki/Double-precision_floating-point_format) values. We can access the individual elements using curved brackets

In [100]:
a(1)

ans = 3

or provide a range of indices to return multiple values, using the `:` syntax

In [102]:
a(2:3)

ans = 1x2 double
     4     5

which will return all the values of `a` between index `2` and index `3`. 

We can create a column vector by designating new rows with `;`

In [104]:
b = [6; 7; 8]

b = 3x1 double
     6
     7
     8

and, similarly, we can access individual elements using curved brackets

In [106]:
b(1:2)

ans = 2x1 double
     6
     7

### Matrices
A matrix is simply a collection of vectors. We can either think of a matrix as a collection of column-vectors, or as a collection of row-vectors. Either way, a matrix has multiple rows *and* multiple columns. For instance

$$
\mathbf{A} = 
\begin{bmatrix}
1 & 2 & 3 \\
4 & 5 & 6 \\
\end{bmatrix}
$$

is a $2 \times 3$ matrix consisting of elements $a_{11} = 1$, $a_{12} = 2$ and so on. We can refer to any element generically as $a_{ij}$ for row $i$ and column $j$. Typically, matrices are written as upper-case **bold** letters, to distinguish them from vectors.

In MATLAB, we declare matrices using square brackets and a combination of spaces and semi-colons. For example, we could declare $\mathbf{A}$ using

In [108]:
A = [1 2 3; 4 5 6]

A = 2x3 double
     1     2     3
     4     5     6

We can also make this declaration look more like a matrix using the MATLAB line-continuation syntax

In [110]:
A = [1 2 3; ...
     4 5 6];

## Matrix Operations
Beyond being containers for collections of numbers, matrices and vectors are *mathematical objects*, that we can treat in similar fashion to regular numbers. As such, arithmetic operations such as addition or subtraction all have matrix equivalents. However, these equivalents may not be particularly intuitive, as we will see below.

### Addition and Subtraction
Starting with addition and subtraction, two matrices can be added or subtracted if they are the same dimensions. In this case, the operation is defined as the addition or subtraction of each individual element. For instance, for a matrix with $m$ rows and $n$ columns, we can write that

$$
\mathbf{C} = \mathbf{A} + \mathbf{B} =
\begin{bmatrix}
a_{11} + b_{11} & a_{12} + b_{12} & \dots  & a_{1n} + b_{1n} \\
a_{21} + b_{21} & a_{22} + b_{22} & \dots  & a_{2n} + b_{2n} \\
\vdots          & \vdots          & \ddots & \vdots          \\
a_{m1} + b_{m1} & a_{m2} + b_{m2} & \dots  & a_{mn} + b_{mn} \\
\end{bmatrix}
$$

where the ellipses ($\dots$) just mean "continue the pattern along this direction". In this example, the row indices run from $i = 1,\dots,m$ and the column indices run from $j = 1,\dots,n$. We can see this in MATLAB

In [10]:
A = [1 2 3; 4 5 6];
B = [2 3 4; 5 6 7];
C = A + B

C = 2x3 double
     3     5     7
     9    11    13

where the result `C` is the same as

In [13]:
C = [1+2 2+3 3+4; 4+5 5+6 6+7]

C = 2x3 double
     3     5     7
     9    11    13

Matrix subtraction works in much the same way, so we have

$$
\mathbf{C} = \mathbf{A} - \mathbf{B}
$$

which, again, is only defined when $\mathbf{A}$ and $\mathbf{B}$ have the same dimensions. For instance

In [19]:
A = [1 4 6; 3 2 3];
B = [2 3 4; 5 6 7];
C = A - B

C = 2x3 double
    -1     1     2
    -2    -4    -4

If we try to add/subtract matrices of different dimensions MATLAB will throw an error

In [24]:
A = [1 3 4; 5 6 7; 8 9 6];
B = [1 3; 6 5];
C = A + B

Arrays have incompatible sizes for this operation.

Formally, we say that addition or subtraction is *undefined* for matrices of different dimensions. Thus, we can see that the rules that govern what we can and cannot do with matrices are stricter than we may be used to with regular numbers. 

### Multiplication
Multiplication of matrices is also possible, however, this is a more complex operation than addition/subtraction. To begin with, multiplying a matrix by a single number simply *scales* the matrix by that amount. As such, in the world of linear algebra, single numbers are often called *scalars*. For instance

In [29]:
A = [1 2 3; 4 5 6];
B = 2 * A

B = 2x3 double
     2     4     6
     8    10    12

Unfortunately, that is where the simplicity ends. Let us consider the case of multiplying two matrices. For example

$$
\mathbf{A} = 
\begin{bmatrix}
    1 & 2 & 3 \\
    4 & 5 & 6
\end{bmatrix} \quad
\mathbf{B} = 
\begin{bmatrix}
    7  & 8  \\
    9  & 10 \\
    11 & 12
\end{bmatrix}
$$

You may notice that these are different dimensions, which is something that will be explained soon. To begin with, let us see what MATLAB tells us $\mathbf{A} \times \mathbf{B}$ is

In [33]:
A = [1 2 3; 4 5 6];
B = [7 8; 9 10; 11 12];
C = A*B

C = 2x2 double
    58    64
   139   154

You might want to take a moment to ponder this result and see if you can work out how MATLAB got there.

To see how this multiplication works, start with the first *row* of $\mathbf{A}$ and the first *column* of $\mathbf{B}$. Each element in the row must be multiplied by each element in the column and then added together. In this example, we do $(1 \times 7) + (2 \times 9) + (3 \times 11) = 58$. This sum then becomes the result for the first element of the new matrix

$$
\begin{bmatrix}
    1           & 2           & 3           \\
    \phantom{4} & \phantom{5} & \phantom{6} \\
\end{bmatrix} \times
\begin{bmatrix}
    7  & \phantom{8}  \\
    9  & \phantom{10} \\
    11 & \phantom{12} \\
\end{bmatrix} =
\begin{bmatrix}
    58            & \phantom{64}  \\
    \phantom{139} & \phantom{154} \\
\end{bmatrix}.
$$

We then do the same thing with the first *row* of $\mathbf{A}$ and the second *column* of $\mathbf{B}$ 

$$
\begin{bmatrix}
    1           & 2           & 3           \\
    \phantom{4} & \phantom{5} & \phantom{6} \\
\end{bmatrix} \times
\begin{bmatrix}
    \phantom{7}  & 8  \\
    \phantom{9}  & 10 \\
    \phantom{11} & 12 \\
\end{bmatrix} =
\begin{bmatrix}
    58            & 64  \\
    \phantom{139} & \phantom{154} \\
\end{bmatrix},
$$

then the second *row* of $\mathbf{A}$ and the first *column* of $\mathbf{B}$ 

$$
\begin{bmatrix}
    \phantom{1} & \phantom{2} & \phantom{3} \\
    4           & 5           & 6 \\
\end{bmatrix} \times
\begin{bmatrix}
    7  & \phantom{8}  \\
    9  & \phantom{10} \\
    11 & \phantom{12} \\
\end{bmatrix} =
\begin{bmatrix}
    58  & 64  \\
    139 & \phantom{154} \\
\end{bmatrix}.
$$

Finally, we do the second *row* of $\mathbf{A}$ and the second *column* of $\mathbf{B}$ 

$$
\begin{bmatrix}
    \phantom{1} & \phantom{2} & \phantom{3} \\
    4           & 5           & 6 \\
\end{bmatrix} \times
\begin{bmatrix}
    \phantom{7}  & 8  \\
    \phantom{9}  & 10 \\
    \phantom{11} & 12 \\
\end{bmatrix} =
\begin{bmatrix}
    58  & 64  \\
    139 & 154 \\
\end{bmatrix}.
$$

If this is the first time you have come across matrix multiplication, this will seem very strange. For a deeper explanation of why this is the case, have a look at the info box below. However, the important thing to know from a practical perspective is that matrix-matrix, matrix-vector, vector-matrix and vector-vector multiplication are only defined when the number of *columns* in $\mathbf{A}$ matches the number of *rows* in $\mathbf{B}$. In addition, the resultant matrix will always have the same number of *rows* as $\mathbf{A}$ and the same number of *columns* as $\mathbf{B}$.

```{admonition} Advanced: Why is Matrix Multiplication So Strange?
:class: dropdown

To understand why matrix multiplication has the definition that it does, we need to first understand that linear algebra is all about *linear equations*. A linear equation is simply an equation where some numbers are multiplied by other numbers and then added together to form a weighted-sum. For example, the equation below is a linear equation

$$
y = a_{1}x_{1} + a_{2}x_{2} + b
$$

where there are two *variables* ($x_{1}$ and $x_{2}$) and three *coefficients* ($a_{1}$, $a_{2}$ and $b$). Each variable is multiplied by its respective coefficient and then everything is added together to form the value of y. The important point here is that the variables can take on any value (hence they can *vary*), whereas the coefficients are usually fixed.

Linear equations are incredibly useful in many branches of science for describing real-world phenomena. Often *multiple* equations are needed to model complex phenomena, which is known as a *system* of linear equations. These equations share the same coefficients, but have different values for the variables. For example, take the following system of 3 equations

$$
y_{1} &= a_{1}x_{11} + a_{2}x_{12} + b \\
y_{2} &= a_{1}x_{21} + a_{2}x_{22} + b \\
y_{3} &= a_{1}x_{31} + a_{2}x_{32} + b
$$

Notice here that each equation has different $x$ values (indicated by the different indices), but the same $a$ and $b$ values. The point of matrices is that they allow this system to be represented compactly *and* provides methods for solving for $a$ and $b$. To see this, we can write the system as

$$
\begin{bmatrix}
y_{1} \\
y_{2} \\
y_{3}
\end{bmatrix}
=
\begin{bmatrix}
x_{11} & x_{12} & 1 \\
x_{21} & x_{22} & 1 \\
x_{31} & x_{32} & 1 \\
\end{bmatrix}
\begin{bmatrix}
a_{1} \\
a_{2} \\
b
\end{bmatrix},
$$

which can be expressed simply as

$$
\mathbf{y} = \mathbf{X}\mathbf{a}.
$$

If we use the definition of matrix multiplication, notice that this recreates the linear equations that define the system. For instance, the first row of $\mathbf{X}$ gives

$$
\begin{bmatrix}
x_{11}           & x_{12}           & 1           \\
\phantom{x_{21}} & \phantom{x_{22}} & \phantom{1} \\
\phantom{x_{31}} & \phantom{x_{32}} & \phantom{1} \\
\end{bmatrix}
\begin{bmatrix}
a_{1} \\
a_{2} \\
b
\end{bmatrix} = 
\begin{bmatrix}
x_{11}a_{1} + x_{12}a_{2} + b \\
\phantom{} \\
\phantom{}
\end{bmatrix}.
$$

You can try this yourself for the remaining two rows. The important point here is that this only works because the rules of matrix multiplication make it work. So, the only way for a system of linear equations to be represented in matrices is by having a more convoluted definition of how multiplication works.

```

### Transposition
As we have seen above, the dimensions of a matrix or vector are very important for determining when matrix multiplication can be performed. In practise, this means that the dimensions of matrices may need to be manipulated. One common approach is an operation called *transposition*, where the rows and the columns of a matrix are swapped. The notation for this is either $\mathbf{M}^{\prime}$ or $\mathbf{M}^{T}$ . So that if

$$
\mathbf{M} = 
\begin{bmatrix}
1 & 2 \\
3 & 4 \\
5 & 6
\end{bmatrix},
$$

the transpose would be

$$
\mathbf{M}^{\prime} = 
\begin{bmatrix}
1 & 3 & 5 \\
2 & 4 & 6
\end{bmatrix}.
$$

You can do this in MATLAB using the `'` symbol

In [62]:
M = [1 2; 3 4; 5 6]

M = 3x2 double
     1     2
     3     4
     5     6

In [64]:
M'

ans = 2x3 double
     1     3     5
     2     4     6

### Division
We have now covered matrix addition, subtraction and multiplication. So what about matrix division? Unfortunately, this is where things get complicated again. To start with, let us review a particular way of viewing division of normal numbers. As we should all know, division is simply splitting a number into equal parts. However, we can also view this as a *multiplication*. Specifically, multiplication by the *inverse* of another number. For instance

$$
10 \times \frac{1}{2} = 5
$$

because

$$
10 \times \frac{1}{2} = \frac{10}{1} \times \frac{1}{2} = \frac{10}{2} = 5.
$$

This is known as the *multiplicative inverse* of a number. Formally, the inverse is whatever number can be multiplied by the original number to make 1. For instance, the multiplicative inverse of 2 is $\sfrac{1}{2}$ because $2 \times \sfrac{1}{2} = 1$. This is written as

$$
2^{-1} = \frac{1}{2}
$$

This is important because there is no definition of division for matrices. However, we know there is multiplication. As such, we perform division by multiplying one matrix by the *inverse* of another.

A special matrix is the *identity* matrix.

## Some Additional Programming Concepts

### Scripts

### Functions
Beyond just writing code within a separate script, we can further containerise MATLAB code by creating a *function*. This is important because SPM is effectively just hundreds of MATLAB functions that can be called in different ways to process and analyse images.

## Summary
In this lesson, we have treated MATLAB as simply a means of understanding some basic concepts in linear algebra. However, MATLAB is really a fully-fledged programming language. We have not touched any programming concepts beyond simply variable assignment, however, if you want to make the most of MATLAB as a tool then you will need to learn to use it as a programming language. Although not an essential part of this course, this would be a skill that will come in very handy in Semester 2, as well as in your dissertation project work. There are many resources available to learn to programme in MATLAB and we would encourage everyone to try and go through them as part of the additional reading throughout this course. Some resources we recommend are:

-
-
-

## Variables
The most basic element of any programming language is the ability to store things in memory for later use. We do this by using a variable, which you can think of as a box with a name that we put something inside to keep for later. As an example, the code below shows the declaration of a variable called a and assigning it the number 5.

In [4]:
a = 5;

This is like having a box called a that contains the number 5. Now, every time we type a, MATLAB will replace it with the number 5. This means we can do the following

In [5]:
a + 7

ans = 12

Similarly, we could declare another variable with a different name.

In [6]:
b = 10;

and then add the variables together

In [7]:
a + b

ans = 15

Although these examples have used numbers, variables can hold any data type that MATLAB recognises. For instance, one of the most useful data types we can save to a variable is text

In [8]:
hello_text = 'hello';

which we could use to store a file path or file name, for instance.

There are a few rules you have to adhere to when naming variables. A variable name cannot start with a number, it cannot contain spaces and the only special character you can use is an underscore. You also need to be aware that MATLAB is case-sensitive, meaning that `hello_text` and `HELLO_TEXT` will be interpreted as different variables. In addition, remember that a variable can only hold one thing at a time and if you re-assign a variable then the original value will be lost.

## Scripts
Typically, if we are writing code to complete some task we will not enter it line-by-line at the MATLAB prompt. Instead, we will write the steps inside a script. You can think of a script as a recipe that we write in code inside a text file. It has multiple steps that are written on different lines that MATLAB will execute one at a time. These scripts can be saved so they can be re-used, or used as templates for other scripts in the future. Indeed, writing scripts is one of the best ways of leaving a paper trail for your analysis, to fully document what you did to the data.

In order to start writing a script click the New Script icon in the MATLAB toolbar to open the blank script editor. We can now enter all our code in this new window. When we want to run it, we save the file and then click the big green Run button in the toolbar. Usefully, we can also include comments in a script. These are pieces of text that MATLAB will ignore, but serve as notes for our future selves. You can include a comment in a script by typing the percentage symbol, followed by whatever text you like. These will be coloured green to make the comments obvious. For example

In [9]:
% This is a comment
a = 5;

% This is another comment. It will be ignored by MATLAB when the script is executed.
b = 10;

% Comments can be used to describe what code is doing. For instance, add a to b and save to c.
c = a + b;

## Control Flow
In the process of writing a script, we may need to decide whether to run code based on some criteria. Alternatively, we may need to repeat a chunk of code a certain number of times. Both of these tasks are examples of control flow where we dictate a more complex path for the code to follow (rather than just running line-by-line). We can break the main types of control flow down into if-statements and for-loops.

### If-statement
An if-statement allows us to make a choice about whether some code is run or not. This is intimately tied to the ideas of variables. Because we do not necessarily know what value a variable will have when we run some code, we may want to check what the value is and then decide what we want to do. An if-statement allows us to ask a question and then decide whether to run some code or not, depending on the answer. This question has to either be true or false, so we can only ask things like "is this variable larger than 5". We can then decide what we want MATLAB to do if the question is true and what we want MATLAB to do if the question is false.

An if-statement has the following structure

In [10]:
if % Ask question
    % What to do if question is true
else
    % What to do if question is false
end

So, for example, if we wanted to set some text to be different if the variable a was larger than 5 compared to if it is smaller than (or equal) to 5, we could write

In [11]:
number  = 20;
message = '';

if number > 5
    message = 'The number you typed is greater than 5';
else
    message = 'The number you typed is is less than (or equal) to 5';
end

disp(message);

The number you typed is greater than 5


When this script is run, you will be prompted to type in a number and then you will see a message about whether the number was greater than 5 or not. Not exactly Nobel prize stuff, but hopefully the use of the if-statement is clear.

We can also place an if-statement inside another if-statement to make different choices if the original question is true. This is known as nesting statements.

In [15]:
number  = 5;
message = '';

if number >= 5
    if number == 5
        message = 'The number you typed was 5';
    else
        message = 'The number you typed was greater than 5';
    end
else
    message = 'The number you typed was less than 5';
end

disp(message);

The number you typed was 5


`````{admonition} Tip
:class: tip
In all of these examples, the symbols `>`, `>=` and `==` are examples of *relational operators* that allow the comparison of variables in different ways. You can read about the operations that MATLAB supports in the [documentation](https://uk.mathworks.com/help/matlab/relational-operators.html).
`````

### For-loops
Beyond checking variable values, we often want to repeat some code a certain number of times in a script. For instance, we may want to copy multiple files, or run some image processing on multiple images. In both cases we want MATLAB to repeat the same set of steps multiple times, but with a different input. For this purpose, all programming languages contain the concept of loops, which allow us to write a set of steps once and then ask the computer to repeat those steps as many times as we want.

The main type of loop we use is known as a for-loop, which has the following structure

In [16]:
for index_variable = start_value:end_value
    % What to do on each iteration
end

and could be used to sum all the number from 1 to 10

In [17]:
number = 10;
sum    = 0;

for i = 1:number
    sum = sum + i;
end

disp(sum)

    55



An important part of the for-loop is the index variable which tells us the current iteration of the loop. In the example above, the index variable is called i. We asked MATLAB to start i with a value of 1 and then continue adding 1 until it reached the value of number, at which point the loop will exit. This is all given in the syntax for i = 1:number. On the first iteration of the loop, i will have a value of 1, on the second iteration i will have a value of 2 and so on. The utility of this is that we can use i inside the loop to know which iteration is currently running. In the example above, we used this to add the current number onto the previous sum, but the real utility of this is that we can use i to select different variables for each iteration of the loop. For instance, if we had a list of images we wanted to process we can use a for-loop and then use the index variable to select the correct image on each iteration. 

For-loops can also be nested, which we will see quite often because these allow us to loop over 2D and 3D structures. For instance, if we wanted to access every voxel of a 3D image we could use three nested for-loops that move over the rows, columns and slices of an image

In [18]:
image_rows = 100;
image_cols = 100;
image_slcs = 100;

for i = 1:image_rows
    for j = 1:image_cols
        for k = 1:image_slcs
            % Do something with the voxel at row i column j slice k
        end
    end
end

## Data Structures
We have now seen how we can use variables and control flow to build basic scripts. However, one limitation of what we have so far is that if we want to access and process many values within the script we would need to assign each to a different variable with a different name. For instance, if we have 100 values we wanted access to, it would be incredibly tedious to assign each one to a separate variable. In these instances, we can make use of various data structures in MATLAB that allow us to assign multiple values to the same variable.

### Matrix

The most basic form of data structure in MATLAB is a matrix. We will talk about the mathematical interpretation of a matrix in the next section, but for now we can just treat it as a container for numbers that has a certain number of rows and a certain number of columns (like a spreadsheet or a table). You specify a matrix in MATLAB using square brackets, where a space indicates a new column and a semi-colon indicates a new row. For example, the following code

In [19]:
A = [1 2 3; 4 5 6]

A = 2x3 double
     1     2     3
     4     5     6

specifies a matrix called A that has 2 rows, 3 columns and contains the numbers 1-6.

Accessing the individual elements of a matrix requires indexing. We use curved brackets after the variable name and then indicate which row and column we want, separated by commas. For example, to access the number 2 we would use the following code to indicate that we want row 1 and column 2

In [20]:
A(1,2)

ans = 2

The utility of the for-loop is hopefully a bit clearer now, as we can use it to access all the elements inside a matrix using the index variables

In [21]:
for i = 1:size(A,1)
    for j = 1:size(A,2)
        number = A(i,j);
        disp(number);
    end
end

     1

     2

     3

     4

     5

     6



where `size(A,1)` tells you how many rows `A` has, `size(A,2)` tells you how many columns `A` has and the `disp()` function tells MATLAB to print the number at the prompt.

### Character Array

Beyond numbers, we may also want to store text in some sort of data structure. This is where the character array comes in. We actually saw a character array earlier when we specified the `hello_text` variable. Rather than being a variable with a single value, what was returned was actually a character array with 1 row and as many columns as letters. So a character array is really the same as a matrix, but the rows and columns refer to letters instead of numbers. A limitation to this approach, however, is that if you want multiple rows there needs to be the same number of characters in each row.

In [22]:
a = ['hello this is some text'; 'hello this is more text']

a = 2x23 char array
    'hello this is some text'
    'hello this is more text'

Alternatively, you have to pad the rows with spaces to make up the difference.

In [23]:
a = ['hello this is some text'; 'hello this is padding  ']

a = 2x23 char array
    'hello this is some text'
    'hello this is padding  '

### Cell Array
The fact that a matrix and a character array behave so similarly is a testament to the MATLAB design principles of treating everything generically like a matrix. The issue with this is that a matrix and a character array can only contain one type of data. This means you need to have multiple variables if you want to store both text and numeric data. To get around this problem the concept of a cell was created. A cell is a container that can hold any type of data. So a cell could contain a matrix, a character array or even another cell. You can then have multiple cells stored together in a single variable called a cell array. 

We can create an empty cell array with 2 rows and 3 columns using the following code

In [25]:
a = cell(2,3)

a = 2x3 cell array
    {0x0 double}    {0x0 double}    {0x0 double}
    {0x0 double}    {0x0 double}    {0x0 double}

Which we could then populate with the names, age and favourite cheese of staff on the course

In [49]:
a{1,1} = 'Martyn';
a{1,2} = 35;
a{1,3} = 'Red Leicester';
a{2,1} = 'Dan';
a{2,2} = 93;
a{2,3} = 'Gorgonzola';
a

a = 2x3 cell array
    {'Martyn'}    {[35]}    {'Red Leicester'}
    {'Dan'   }    {[93]}    {'Gorgonzola'   }

So now we have a single variable called `a` that contains both numbers and text. Notice as well that we can have text in multiple rows without constraints on its length. We can then use the same indexing approaches used previously to access the data

In [50]:
a{2,1}
a{2,2}
a{2,3}

ans = 'Dan'

ans = 93

ans = 'Gorgonzola'

`````{admonition} Tip
:class: tip
Notice that we index a cell array using curly brackets, rather than curved brackets. The distinction is that curved brackets would return the cell itself, whereas curly brackets would return the contents of the cell. This is a confusing design decision on the part of MathWorks, so do not worry if this is not that clear.
`````

### Struct

Continuing the theme of data structures that can store different types, an alternative to a cell array is to use something called a struct. Where a struct differs from a cell array is that rather than using indices a struct uses named fields to store the data. This is useful because we can access data by name rather than having to remember what index we need for each type of data. 

Structs are defined by specifying the name of the struct and then using a dot to create fields. For instance

In [53]:
Lecturer.Name   = 'Martyn';
Lecturer.Age    = 35;
Lecturer.Cheese = 'Red Leicester';

Now we have a struct called `Lecturer` with the fields `Name`, `Age` and `Cheese`. We can then access the values of the fields using the same dot syntax.

In [54]:
Lecturer.Cheese

ans = 'Red Leicester'

Structs can also be nested, which means you can end up with some quite complex multi-level structures

In [55]:
clear Lecturer;

Lecturer.Name.Title = 'Dr';
Lecturer.Name.First = 'Martyn';
Lecturer.Name.Last  = 'McFarquhar';

So here the struct `Lecturer` contains another struct called `Name` which has the fields `Title`, `First` and `Last`. A trick you can use with an unfamiliar struct is to type the name of the struct, then a dot and then press the <TAB> key to see a menu pop-up of the fields that are available.

We can also combine indices and fields to create an array of structs. For example

In [59]:
Lecturer(1).Name.Title  = 'Dr';
Lecturer(1).Name.First  = 'Martyn';
Lecturer(1).Name.Last   = 'McFarquhar';
Lecturer(1).Age         = 35;
Lecturer(1).Cheese      = 'Red Leicester';

Lecturer(2).Name.Title  = 'Lord High Admiral';
Lecturer(2).Name.First  = 'Daniel';
Lecturer(2).Name.Last   = 'Cox';
Lecturer(2).Age         = 93;
Lecturer(2).Cheese      = 'Gorgonzola';

which we can then index in the usual fashion

In [57]:
Lecturer(2)

ans = struct with fields:
      Name: [1x1 struct]
       Age: 93
    Cheese: 'Gorgonzola'

In [60]:
Lecturer(2).Name

ans = struct with fields:
    Title: 'Lord High Admiral'
    First: 'Daniel'
     Last: 'Cox'

## Functions

The final topic of this section is one of the most important concepts that we need for programming with MATLAB. One of the central tenets of writing code that can be re-used and easily maintained is to make it modular. This means breaking the code up into independent chunks that do very specific jobs. Those individual chunks are written to be very generic so they can be re-used in many places. The way we achieve this is by isolating chunks of code into functions. You can think of these like a black box that we pass data into and it gives us some data back again.

An example of a function is given below

In [43]:
function a = add_five(b)
    a = b + 5;
end

This declares a function called `add_five` that takes a single input variable called `b` and returns a single output variable called `a`. From the body of the function, we can see that all this does is take the value of `b` and adds five to it.

In order to use this function we have to write it in a script and then save it using the same filename as the function. So in this case we would save the script as `add_five.m`. This needs to be saved in a location that is on the MATLAB path, so either the current directory or Documents > MATLAB are safe bets. Once you have done that, you can use the `add_five` function at the prompt or in another script like so

In [44]:
add_five(7)

ans = 12

So what we have done here is saved code in a format where we can use it again simply by typing `add_five` with whatever argument we like. Now, obviously, a function that just adds 5 to a number is a bit useless. However, if you think of something more complex, such as reading neuroimaging data from a file name and returning a matrix of voxel values, you can start appreciating the power of functions. Not only do they make code re-useable, but they also make code easier to understand. For instance, seeing `image = read_data('image_file.nii')` in a script is a lot easier to understand than suddenly seeing 300 lines of code for reading the file from disk. We will see functions come up a lot during this course, mainly in terms of the functions that come with SPM. The main thing to understand is that a function is effectively hiding code from you and that in order to use a function you just need understand what it needs in terms of input variables and what it will return in terms of output variables.

## Summary

It may seem unfair to call this section a brief introduction to programming in MATLAB, however, the world of computer programming is huge and we have only covered the very basics here. If you are new to programming, this may have felt a bit overwhelming, but do not worry. As we said at the outset, treat this section as a reference that you can come back to in future. A lot of these topics will become clearer over time as you see more examples of code and scripts. What we want at this stage is for you to have some familiarity with this so that the first time you see a MATLAB script you are not doing so cold. We will be providing more guidance on this as we go through the course. If this has whetted your appetite and you want to learn more, there are many books on MATLAB programming, such as MATLAB for Psychologists by Borgo, Soranzo and Grassi.